src/cascadia/TerminalCore/TerminalSelection.cpp

```diff
@@ -235,92 +73,173 @@ const bool Terminal::IsCopyOnSelectActive() const noexcept
+    _selectionStart = _selectionPivot;
+
+    SetSelectionEnd(viewportPos);
 }
 
 // Method Description:
-// - Record the position of the beginning of a selection
+// - Update selection anchors when dragging to a position
+// - based on the selection expansion mode
 // Arguments:
-// - position: the (x,y) coordinate on the visible viewport
-void Terminal::SetSelectionAnchor(const COORD position)
+// - viewportPos: the (x,y) coordinate on the visible viewport
+// - expansionModeOverwrite: overwrites the _multiClickSelectionMode for this function call. Used for ShiftClick
+void Terminal::SetSelectionEnd(const COORD viewportPos, std::optional<SelectionExpansionMode> expansionModeOverwrite)
 {
-    _selectionAnchor = position;
+    auto textBufferPos = _ConvertToBufferCell(viewportPos);
+    _buffer->GetSize().Clamp(textBufferPos);
 
-    // include _scrollOffset here to ensure this maps to the right spot of the original viewport
-    THROW_IF_FAILED(ShortSub(_selectionAnchor.Y, gsl::narrow<SHORT>(_scrollOffset), &_selectionAnchor.Y));
+    // if this is a shiftClick action, we need to overwrite the _multiClickSelectionMode value (even if it's the same)
+    // Otherwise, we may accidentally expand during other selection-based actions
+    const auto shiftClick = expansionModeOverwrite.has_value();
+    _multiClickSelectionMode = expansionModeOverwrite.has_value() ? *expansionModeOverwrite : _multiClickSelectionMode;
 
-    // copy value of ViewStartIndex to support scrolling
-    // and update on new buffer output (used in _GetSelectionRects())
-    _selectionVerticalOffset = gsl::narrow<SHORT>(ViewStartIndex());
+    switch (_multiClickSelectionMode)
+    {
+    case SelectionExpansionMode::Line:
+        _ChunkSelectionByLine(textBufferPos, shiftClick);
+        break;
+    case SelectionExpansionMode::Word:
+        _ChunkSelectionByWord(textBufferPos, shiftClick);
+        break;
+    case SelectionExpansionMode::Cell:
+    default:
+        _ChunkSelectionByCell(textBufferPos, shiftClick);
+
+        if (_copyOnSelect && !_IsSingleCellSelection())
+        {
+            _allowSingleCharSelection = true;
+        }
+        break;
+    }
 
     _selectionActive = true;
-    _allowSingleCharSelection = (_copyOnSelect) ? false : true;
-
-    SetEndSelectionPosition(position);
+}
 
-    _multiClickSelectionMode = SelectionExpansionMode::Cell;
+// Method Description:
+// - Update the selection anchors when dragging to a position
+// - Used for SelectionExpansionMode::Word
+// Arguments:
+// - targetPos: the (x,y) coordinate we are dragging to on the text buffer
+// - shiftClick: when enabled, only update the _endSelectionPosition
+void Terminal::_ChunkSelectionByCell(const COORD targetPos, bool /*shiftClick*/)
+{
+    // TODO GH #4557: Shift+Multi-Click Expansion
+    //   shiftClick is reserved for that implementation
+    if (_buffer->GetSize().CompareInBounds(targetPos, _selectionPivot) <= 0)

```

Comment:
all three of these functions do the same exact thing. How about this?

```
std::tuple<COORD, COORD> _AdjustStartAndEndAcrossPivot(const COORD targetPos, const COORD endpoint1, const COORD endpoint2);

...


std::tie(_selectionStart, _selectionEnd) = AdjustThingByThing(targetPos, Expand(targetPos), Expand(_selectionPivot))
```

but maybe the complexity of the args we're passing to AdjustThingByThing makes it reasonable to just copy the logic here for clarity...